### ISSUE TITLE
[MCOMPILER-534] Document conditional setting of the --release property

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-compiler-plugin#756]
**[Gary D. Gregory](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=garydgregory)** opened **[MCOMPILER-534](https://issues.apache.org/jira/browse/MCOMPILER-534?redirect=false)** and commented

Apache Commons Parent 57 RC1 (https://lists.apache.org/thread/rwgf9c7dgl2bymq3jg6ckxs6gk9pz4mg) attempts to use the javac release flag only on Java 9 and above with 

```xml
<maven.compiler.release>8</maven.compiler.release>
```

and

```xml
<profile>
  <id>jdk9-compiler</id>
  <activation>
    <jdk>[9</jdk>
  </activation>
  <build>
         <pluginManagement>
          <plugins>
        <plugin>
          <groupId>org.apache.maven.plugins</groupId>
          <artifactId>maven-compiler-plugin</artifactId>
          <configuration>
            <release>${maven.compiler.release}</release>
          </configuration>
        </plugin>
      </plugins>
    </pluginManagement>
  </build>
</profile>
```

This fails miserably:

```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  5.402 s
[INFO] Finished at: 2023-04-23T09:23:21-04:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.11.0:compile (default-compile) on project commons-lang3: Fatal error compiling: invalid flag: --release -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

IMO, setting a property "maven.compiler.release" should not be used on compilers that do not support the flag.

---

**Affects:** 3.11.0

**Remote Links:**
- [GitHub Pull Request #187
](https://github.com/apache/maven-compiler-plugin/pull/187)

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

You can't do this. You profile is moot since you have set the property already globally. It is passed directly to the plugin. Either set the property within the profile or set a literal value.

**Comment by jira-importer** [comment]
**[Alex Herbert](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=alexherbert)** commented

We know the supplied configuration is incorrect for conditional setting of the property. The property is for targeting Java 8 when building with a later JDK. But you do not have to set the property if you are building with JDK 8. If you do then you receive the described error.

Perhaps the conditional inclusion of the property (for JDK 9+) can be added to the example on the maven-compiler-plugin documentation:

[Setting the --release of the Java Compiler](https://maven.apache.org/plugins/maven-compiler-plugin/examples/set-compiler-release.html)

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

PR for doc improvements welcome!

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[ggregory](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ggregory), you can introduce a `javaVersion` property just like we do with Maven parent and use that value everywhere. For release, in the profile only of course. No need to fiddle with the plugin, it should pick it up automatically. Give it a try.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [6dfe7f9708a329c51005b79976574c090ca43655](https://gitbox.apache.org/repos/asf?p=maven-compiler-plugin.git;a=commit;h=6dfe7f9708a329c51005b79976574c090ca43655).
